src/field/FieldDriver.i.hh

```diff
@@ -0,0 +1,269 @@
+ */
+CELER_FUNCTION auto
+FieldDriver::find_next_chord(real_type step, const OdeState& state)
+    -> FieldOutput
+{
+    // Output with a step control error
+    FieldOutput output;
+
+    // Try with the proposed step
+    output.step_taken = step;
+
+    bool          converged       = false;
+    unsigned int  remaining_steps = shared_.max_nsteps;
+    StepperResult result;
+
+    do
+    {
+        result = stepper_(step, state);
+
+        // Check whether the distance to the chord is small than the reference
+        real_type dchord
+            = distance_chord(state, result.mid_state, result.end_state);
+
+        if (dchord <= shared_.delta_chord + rel_tolerance())
+        {
+            converged    = true;
+            output.value = truncation_error(
+                step, shared_.epsilon_rel_max, state, result.err_state);
+        }
+        else
+        {
+            // Estimate a new trial chord with a relative scale
+            output.step_taken
+                *= std::fmax(std::sqrt(shared_.delta_chord / dchord), half());
+        }
+    } while (!converged && (--remaining_steps > 0));
+
+    CELER_ASSERT(converged);
+
+    // Update new position and momentum
+    output.state = result.end_state;
+
+    return output;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Accurate_advance for an adaptive step control: Perform an adaptive step
+ * integration for a proposed step or a series of sub-steps within a required
+ * tolerance until the the accumulated curved path is equal to the input step
+ * length.
+ */
+CELER_FUNCTION real_type FieldDriver::accurate_advance(real_type step,
+                                                       OdeState* state,
+                                                       real_type hinitial)
+{
+    CELER_ASSERT(step > 0);
+
+    // Set an initial proposed step and evaluate the minimum threshold
+    real_type end_curve_length = step;
+
+    real_type h = ((hinitial > rel_tolerance() * step) && (hinitial < step))
+                      ? hinitial
+                      : step;
+    real_type h_threshold = shared_.epsilon_step * step;
+
+    // Output with the next good step
+    FieldOutput output;
+
+    // Performance integration
+    bool         succeeded       = false;
+    real_type    curve_length    = 0;
+    unsigned int remaining_steps = shared_.max_nsteps;
+
+    do
+    {
+        output = this->integrate_step(h, *state);
+
+        curve_length += output.step_taken;
+
+        if (h < h_threshold || curve_length >= end_curve_length)
+        {
+            succeeded = true;
+        }
+        else
+        {
+            h = std::fmax(output.value, shared_.minimum_step);
+            if (curve_length + h > end_curve_length)
+            {
+                h = end_curve_length - curve_length;
+            }
+        }
+        *state = output.state;
+    } while (!succeeded && --remaining_steps > 0);
```

Comment:
Similar remark as for `FieldDriver::find_next_chord`

Reply:
Similar loop check applied for find_next_chord.